mitodl/mitxonline
# feat: add course run upgrade deadline

#### Pre-Flight checklist


#### What are the relevant tickets?
#792 

#### What's this PR do?
- Adds upgrade deadline to a course run
- Manages the enroll visibility based on upgrade deadline
- Manages cart checkout w.r.t upgrade deadline
- Manages Dashboard enrollments w.r.t upgrade deadline
- Manages course detail visibility on the basis of upgrade deadline
- Removed unused imports from some frontend js file

#### How should this be manually tested?

There are a couple of cases this should be tested with But also take a look at the communication on the ticket:
- Make sure to have courses sets with & without products
- Setup on this branch and without doing anything this should work the same as above (In this case the upgrade_deadline will be empty so it should behave the same for courses with/without product)
- Set an upgrade deadline in the future for the course run, Open the course detail page, and enroll. (The enroll button should work fine as it used to e.g. A modal in case of a product and directly enroll in case of no product)
- Now set the upgrade_deadline for the course run in past, The enroll button should not show the Upgrade modal.
- Enroll in the course (free mode), Visit Dashboard, and see you don't see the `Get Certificate` button there as well
- Visit the course detail page(of the above-enrolled course) and check that you see `Get Certificate` card.
- Now set a future upgrade_deadline and add the course to the cart, now change the upgrade deadline in past and try to checkout, it should show an error that a course in the cart can not be upgraded
- If you think of more scenarios to check this with (try them)

- Would be good to give a quick look at removed imports in frontend/js code to make sure they were really unused

#### Screenshots (if appropriate)

**COURSE UPGRADE DEADLINE IN PAST**


**When a user tries to checkout with a course with upgrade_deadline in the past**
<img width="1680" alt="Screenshot 2022-08-15 at 5 10 37 PM" src="https://user-images.githubusercontent.com/34372316/184632805-d4797aaf-38d4-44a4-89b9-39bb4fc2f754.png">

**When a user is enrolled in a course with valid product but the upgrade_deadline is in past**
<img width="1669" alt="Screenshot 2022-08-15 at 5 10 12 PM" src="https://user-images.githubusercontent.com/34372316/184632870-de58724e-013c-45b1-932a-6038c27ba940.png">

**When a user is enrolled in a course free mode with a product but upgrade_deadlin is in past (He doesn't see `Get Certificate` button the dashboard)**
<img width="1532" alt="Screenshot 2022-08-15 at 5 10 21 PM" src="https://user-images.githubusercontent.com/34372316/184632992-4f88f87d-091f-4a20-9a1c-16bca2ee4bf8.png">


**COURSE UPGRADE DEADLINE IS EMPTY OR IN THE FUTURE**

**Course Detail Page**
<img width="1564" alt="Screenshot 2022-08-15 at 5 17 37 PM" src="https://user-images.githubusercontent.com/34372316/184633909-aeb26846-0772-4d7a-be42-713afa286e3b.png">

**Dashboard Page**
<img width="1512" alt="Screenshot 2022-08-15 at 5 17 45 PM" src="https://user-images.githubusercontent.com/34372316/184633920-b6c0aaca-5148-45cb-9cc6-76244ebd1e6c.png">

**Enroll button click**
<img width="1486" alt="Screenshot 2022-08-15 at 5 18 05 PM" src="https://user-images.githubusercontent.com/34372316/184633922-262c8470-264f-4806-a3ff-022f5b3236a9.png">
